Add host filters

jakajancar opened the pull request:
Resolves issue https://github.com/seanmonstar/warp/issues/432.
Obsoletes PR https://github.com/seanmonstar/warp/pull/489 (better logic, added tests and documentation).

cc @dmeijboom

@jxs, I'd appreciate a review.

dmeijboom commented:
Nice work 👍🏽 I closed my PR in favor of this one.

seanmonstar commented on src/filters/authority.rs:
````diff
@@ -0,0 +1,92 @@
+///     warp::authority("foo.com").map(|| "you've reached foo.com")
+///     .or(warp::authority("bar.com").map(|| "you've reached bar.com"));
+/// ```
+pub fn authority(expected: &str) -> impl Filter<Extract = (), Error = Rejection> + Clone {
````
Since this filter requires a match, perhaps the name should indicate that. What these options: `required`, `guard`, or `exact` (also used in `warp::header::exact`)?

jakajancar replied:
Ok, I'll rename it to `exact` to match `header::exact` and remove the `warp::authority` altogether (so you either have to use `::exact` or `::optional`)

seanmonstar commented on src/filters/authority.rs:
```diff
@@ -0,0 +1,92 @@
+                    .and_then(|value| Authority::from_str(value).map_err(|_| reject::invalid_header(name)))
+            );
+
+        future::ready(match (from_uri, from_header) {
```
Since checking the headers requires a hashmap lookup, I'd check `from_uri` first, and only if `None`, *then* try to get the value from the headers.

jakajancar replied:
I kinda like some strictness and see it fail in case there is a discrepancy between URL and Host, rather than just use one of them.

I would also imagine that in the overall inefficient design of warp where you need to do `::exact` multiple times in subtrees (as opposed to `::optional` once and route to a sub-filter dynamically based on that), a hash lookup is a comparatively tiny inefficiency.

So I would leave it as it is, unless you insist.

seanmonstar replied:
What you say makes sense, I didn't notice the subtlety in the implementation.

jakajancar commented:
> I initially read "authority" as "authorization". I wonder if others would think similarly. What about using the term "host" instead?

I was thinking about that. In the end, I went with Authority because:

 1. hyper and the http crate call it authority (and the struct you get is `http::uri::Authority`),
 2. HTTP2 and 3 normally refer to it as (server) authority, and the `Host` header is more of a historic thing.

But absolutely agree re. familiarity and happy to change.  I would keep using `http::uri::Authority` though, but let me know.

jakajancar commented:
@seanmonstar Can you please provide answers/acknowledge that my suggestions are OK on my 2 questions?

seanmonstar commented:
I agree that the header host is phased out for `:authority`... I suppose I'd wonder what other frameworks expose. If everything at the framework level calls it host, does that mean we should also? I don't think we have to. But maybe it's less confusing to users?

jakajancar commented:
- Renamed filter from `authority` to `host`
  - Removed "top-level" filter in favor of explicit `warp::host::exact()` and `warp::host::optional()`

"Host" does seem much more familiar. The returned object is still `Authority` (from `http::uri::Authority`), but hopefully the documentation bridges between the familiar & easy-to-find and the technically accurate.
